### PR TITLE
[android] apply Network Security Config file (fixes #22375) (part 2 of #23105)

### LINKED ISSUE #22375: [Android] Android 9 can not access bundler via cleartext request without additional config
## Environment
```
  React Native Environment Info:
    System:
      OS: Linux 4.18 Arch Linux undefined
      CPU: x64 AMD Ryzen 7 1700X Eight-Core Processor
      Memory: 682.77 MB / 7.79 GB
      Shell: 5.6.2 - /bin/zsh
    Binaries:
      Node: 10.7.0 - /usr/local/bin/node
      Yarn: 1.12.3 - /usr/local/bin/yarn
      npm: 6.4.1 - /usr/local/bin/npm
      Watchman: 4.7.0 - /usr/local/bin/watchman
    SDKs:
      Android SDK:
        Build Tools: 23.0.1, 23.0.3, 25.0.1, 25.0.2, 25.0.3, 26.0.1, 26.0.2, 26.0.3, 27.0.2, 27.0.3, 28.0.1, 28.0.3
        API Levels: 23, 24, 26, 27, 28
    npmPackages:
      react-native: 0.57.x => 0.57.4 
    npmGlobalPackages:
      create-react-native-app: 1.0.0
      react-native-git-upgrade: 0.2.7
```

## Description
As of Android  9 (API Level 28), the default behavior is to now [block all clear text requests](https://developer.android.com/training/articles/security-config#CleartextTrafficPermitted), unless a domain is specifically configured to allow this behavior. As a result the bundler URL is not reachable in development mode, with a generic error that offers no context to what the problem could be, or what solutions are available.

The issue is fairly trivial to resolve once the root cause is apparent, but getting to that root cause can take a bit of searching.

## Reproducible Demo
Start a `react-native` app on an Android 9 devices (I only tested physical device), in development mode. The app will not start, and will complain that the bundle can not be fetched.

### PR DESCRIPTION
**Pull request by Salakar**
This is a follow-up PR for https://github.com/facebook/react-native/pull/23105 - as mentioned on discord.

---

This PR applies the network security config for the RN template project only. New RN projects started with the updated template will be able to connect to the packager on builds built with Android API 28 & above.

See https://developer.android.com/training/articles/security-config#CleartextTrafficPermitted for more information about this newly required config, specifically:

image

Changelog:
----------
[ANDROID] [Template] add Network Security Config file to allow access to packager via cleartext requests in Android API 28 and above. (fixes #22375)

Test Plan:
----------

**Before change**:
image

**After change**:
image

### CODE REVIEW

**Comment by dulmandakh** [comment]
And how about changing https://github.com/facebook/react-native/blob/master/template/android/app/src/main/AndroidManifest.xml? Maybe you can add XML attributes specifying Android API in AndroidManifest.xml

**Comment by hey99xx** [comment]
@Salakar thanks for making progress and also mentioning me. 

Now what you do is use `android:networkSecurityConfig="@xml/rn_network_security_config"` in `ReactAndroid/src/debug/AndroidManifest.xml`. This is not necessarily wrong, but I don't think it will work as you intended.

This is because when you install react-native from NPM, it comes with only one AAR file, the one named react-native-0.56.1.aar somewhere inside in `node_modules/react-native`. It is possible to release two separate AARs for debug and release but it's usually cumbersome for consumers and it sometimes creates issues with transitive dependencies.

I'm not going to suggest editing `template/android/app/src/main/AndroidManifest.xml` either, because that applies to both release and debug APKs and we wouldn't want to expose localhost in production.

Instead you can move
- `ReactAndroid/src/debug/AndroidManifest.xml` into `template/android/app/src/debug/AndroidManifest.xml` and 
- `ReactAndroid/src/main/res/devsupport/xml/network_security_config.xml` into `template/android/app/src/debug/res/xml/network_security_config.xml`

Basically we do the same thing described in https://stackoverflow.com/a/46321866 Note that this solution obviously does not work for brownfield apps that wasn't started from the helloworld template in this repo, so they would have to manually add those 2 files. I think that's acceptable, but I'm not authority here :)

**Comment by Salakar** [comment]
> And how about changing https://github.com/facebook/react-native/blob/master/template/android/app/src/main/AndroidManifest.xml?

Not needed for that one as it's applied as part of `ReactAndroid` build into the final AndroidManifest file that's generated.

**Comment by Salakar** [comment]
Ah damn, @hey99xx you're right again 🙈 this change would only apply when using RN from source like I was locally, completely forgot it gets distributed as an AAR - which this change won't apply to the bundled AAR that gets publish alongside on NPM >.>

I can apply it to the template but like you said this only works for apps that began with this template after the change. Also; is this template moving to the CLI @grabbou?

I may have an alternative though; I can move it to the main manifest and wrap the security config in a `<debug-overrides>` block: https://developer.android.com/training/articles/security-config#debug-overrides - so it'll only apply to debug still

**Comment by hey99xx** [comment]
> I can move it to the main manifest and wrap the security config in a <debug-overrides> block

You can try, but as far as I recall it won't work, I've tried it myself in the past 🙃. `<debug-overrides>` can only be used to override `<trust-anchors>` block inside. I wish it worked according to the common intuition like you have.

Maybe you can ask in the Discord channel for alternative solutions. Unfortunately I don't know a better way. Maybe Facebook is ok putting this in the main AndroidManifest.xml and let anyone worried about security override it, that's another way to handle things.

**Comment by cpojer** [comment]
@Salakar do you have any update on this PR?

**Comment by Salakar** [comment]
@cpojer sorry for the delay getting back to you, 🙈I've been been held up with prior OSS commitments this week - I've not forgotten I owe you a writeup for this as per Discord. 

Will get this done by Friday latest 👍

**Comment by Salakar** [comment]
Hey @cpojer;

I've updated this to apply only to the `template` project. As discussed with @hey99xx above, this can't be applied to RN itself as RN is distributed as a single aar and distributing multiple aars (e.g. debug/release variants) is probably a bit OTT for this issue and could potentially cause more problems than it's worth. 

I've also updated the PR description to reflect the above + added testing screenshots.

I will also follow up with a separate docs PR to https://facebook.github.io/react-native/docs/integration-with-existing-apps - to document this for existing apps that aren't using the template.

**Comment by dulmandakh** [comment]
how about having single config file, but use debug-overrides in it as described in https://developer.android.com/training/articles/security-config. so it'll be empty for non-debug builds.

**Comment by Salakar** [comment]
@dulmandakh see the comments above, specifically;

> I can move it to the main manifest and wrap the security config in a `<debug-overrides>` block

> You can try, but as far as I recall it won't work, I've tried it myself in the past 🙃. <debug-overrides> can only be used to override `<trust-anchors>` block inside. I wish it worked according to the common intuition like you have.
